bank/api/api.go

```diff
@@ -0,0 +1,243 @@
+			return
+		}
+	})
+}
+
+func DeleteUserAccount(accStore domain.AccountStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		a := context.Get(r, "account")
+		account := a.(domain.Account)
+
+		result, err := accStore.GetAccount(account.UserID, account.Currency)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		err = accStore.DeleteAccount(account.UserID, account.Currency)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		err = accStore.InsertHistory(result.AccountID, "account deleted", result.Currency, 0)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	})
+}
+
+func UserAccountWithdraw(accStore domain.AccountStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		a := context.Get(r, "account")
+		account := a.(domain.Account)
+
+		result, err := accStore.GetAccounts(account.UserID)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+
+		for _, current := range *result {
+			if current.Currency == account.Currency {
+				if current.Amount >= account.Amount {
+					withdrawAmt := account.Amount
+					account.Amount = current.Amount - account.Amount
+					err := accStore.UpdateAccount(account)
+					if err != nil {
+						http.Error(w, err.Error(), http.StatusInternalServerError)
+					} else {
+						err = accStore.InsertHistory(current.AccountID, "withdraw", current.Currency, withdrawAmt)
+						if err != nil {
+							http.Error(w, err.Error(), http.StatusInternalServerError)
+						}
+					}
+				} else {
+					http.Error(w, "cannot withdraw more than have", http.StatusBadRequest)
+				}
+			}
+		}
+	})
+}
+
+func UserAccountDeposit(accStore domain.AccountStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		a := context.Get(r, "account")
+		account := a.(domain.Account)
+
+		result, err := accStore.GetAccounts(account.UserID)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+
+		for _, current := range *result {
+			if current.Currency == account.Currency {
+				account.Amount = current.Amount + account.Amount
+				err := accStore.UpdateAccount(account)
+				if err != nil {
+					http.Error(w, err.Error(), http.StatusInternalServerError)
+				} else {
+					err = accStore.InsertHistory(current.AccountID, "deposit", current.Currency, account.Amount-current.Amount)
+					if err != nil {
+						http.Error(w, err.Error(), http.StatusInternalServerError)
+					}
+				}
+			}
+		}
+	})
+}
+
+func UserTransactionHistory(accStore domain.AccountStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		h := context.Get(r, "history")
+		history := h.(domain.History)
+
+		result, err := accStore.GetHistory(history.AccountID)
```

Comment:
Please be sure that accountHistory could be retrieved only for the accounts that are own by user.